src/wallet/bdk_wallet/wallet.rs

```diff
@@ -25,3 +36,191 @@ pub struct Wallet {
+        let _ = index
+            .insert_descriptor(keychain, descriptor.clone())
+            .map_err(|e| match keychain {
+                Keychain::External => {
+                    panic!("this is the first descriptor we're inserting")
+                }
+                Keychain::Internal => {
+                    use bdk_chain::keychain::InsertDescriptorError;
+                    match e {
+                        InsertDescriptorError::DescriptorAlreadyAssigned { .. } => {
+                            DescriptorError::ExternalAndInternalAreTheSame
+                        }
+                        InsertDescriptorError::KeychainAlreadyAssigned { .. } => {
+                            unreachable!("this is the first time we're assigning internal")
+                        }
+                    }
+                }
+                _ => unreachable!("we are not assigning other keychains"),
+            })?;
+
+        // build respective signer container
+        let signer_container = Arc::new(SignersContainer::build(keymap, &descriptor, secp));
+        signers_map.insert(keychain, signer_container);
+    }
+
+    Ok(signers_map)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use bdk_chain::indexed_tx_graph::Indexer;
+    use bip39::Mnemonic;
+    use bitcoind::tempfile::tempdir;
+
+    #[test]
+    fn test_setup_signers_and_update_index() {
+        // Initialise MetaStore
+        let temp_dir = tempdir().unwrap();
+        let file_path = temp_dir.path().join("test_wallet.cbor");
+        let mnemonic = Mnemonic::generate(12).unwrap().to_string();
+
+        let meta_store = MetaStore::init(
+            "test_wallet".to_string(),
+            &file_path,
+            Network::Bitcoin,
+            mnemonic,
+            "passphrase".to_string(),
+            None,
+        )
+        .expect("failed to initialise meta store");
+
+        // Create index and call the function
+        let mut index = KeychainTxOutIndex::<Keychain>::default();
+        let secp = Secp256k1::new();
+
+        let result = initialize_signers_and_update_index(&meta_store, &secp, &mut index)
+            .expect("failed to mutate index & create signers map");
+
+        // Checks for result
+        assert_eq!(result.len(), 2, "Only two keychains key-pair must exist");
+
+        let signer_container_ext = result
+            .get(&Keychain::External)
+            .expect("External keychain must exist");
+        assert_eq!(
+            signer_container_ext.as_key_map(&secp).len(),
+            1,
+            "Descriptor contain 1 xpriv, must have 1 pair for external"
+        );
+
+        let signer_container_int = result
+            .get(&Keychain::Internal)
+            .expect("Internal keychain must exist");
+        assert_eq!(
+            signer_container_int.as_key_map(&secp).len(),
+            1,
+            "Descriptor contain 1 xpriv, must have 1 pair for internal"
+        );
```

Comment:
Instead of just asserting the length of the keychain we can assert the path of the descriptor for external and internal also the fingerprint of the descriptor that it matches with the master key. That will be the right test. Just saying that we have 1 descriptor doesn't tell us anything that we have the right one.